src/simulations/collect.jl

```diff
@@ -343,19 +345,24 @@ function multi_agent_agg_types!(types::Vector{Vector{T} where T}, utypes::Tuple,
 end
 
 """
-    aggname(k) → name
-    aggname(k, agg) → name
-    aggname(k, agg, condition) → name
+    dataname(k) → name
+    dataname(k, agg) → name
+    dataname(k, agg, condition) → name
 
 Return the name of the column of the `i`-th collected data where `k = adata[i]`
 (or `mdata[i]`).
-`aggname` also accepts tuples with aggregate and conditional values.
+`dataname` also accepts tuples with aggregate and conditional values.
 """
-aggname(k, agg) = string(agg)*"_"*string(k)
-aggname(k, agg, condition) = string(agg)*"_"*string(k)*"_"*string(condition)
-aggname(x::Tuple{K,A}) where {K,A} = aggname(x[1], x[2])
-aggname(x::Tuple{K,A,C}) where {K,A,C} = aggname(x[1], x[2], x[3])
-aggname(x::Union{Function, Symbol, String}) = string(x)
+dataname(k, agg) = string(agg)*"_"*string(k)
+dataname(k, agg, condition) = string(agg)*"_"*string(k)*"_"*string(condition)
+dataname(x::Tuple{K,A}) where {K,A} = dataname(x[1], x[2])
+dataname(x::Tuple{K,A,C}) where {K,A,C} = dataname(x[1], x[2], x[3])
+dataname(x::Union{Symbol,String}) = string(x)
+# This takes care to include fieldnames and values in the column name to make column names unique
+# if the same function is used with different values of outer scope variables.
+dataname(x::Function) = join(
+    vcat([string(x)], ["$(prop)=$(getproperty(x, prop))" for prop in propertynames(x)]), "_")
+@deprecate aggname dataname
```

Comment:
I'm not conviced about this part. Is there a reason to make a function `dataname` that accepts both tuples, but also expanded tuples? 

I think `dataname(k)` should be the only valid call signature, and if `k` is a tuple we know what to do internally. It should not be part of the public APi the version `dataname(k, agg, condition)`.

Reply:
I agree, it would make more sense to leave the two and three argument methods out in favor of the tuple version. However, this was what the old `aggname` function did, so I left it in. I'm happy to deprecate the two and three argument `aggname` methods.

For this, do you know how to write a method that accepts an NTuple of length 2 or 3? Right now I have
```julia
dataname(x::NTuple{Union{Symbol,String,Function},2}) = join([dataname(s) for s in x], "_")
dataname(x::NTuple{Union{Symbol,String,Function},3}) = join([dataname(s) for s in x], "_")
```
but it would be nice to merge these two lines.

On the other hand, it might be more readable to keep it as
```julia
dataname(x::Tuple{K,A}) where {K,A} = dataname(x[1])*"_"*dataname(x[2])
dataname(x::Tuple{K,A,C}) where {K,A,C} = dataname(x[1])*"_"*dataname(x[2])*"_"*dataname(x[3])
```
i.e. similar to what the `aggname` code does.

Reply:
Just write `dataname(x::Tuple) = join([dataname(s) for s in x], "_")` ? No need to specialize further on the tuple.

Reply:
Ok, that makes sense.

Reply:
> I'm happy to deprecate the two and three argument aggname methods.

Sorry, missed that, just don't mention them in the docstring, that's enough.

Reply:
Hmmmm... there is an issue with the call semantics. `aggname` was written such that if you have a symbol `:weight` and you aggregate with `mean`, calling `aggname(:weight, mean)` you'll get the column name "mean_weight". With the new calling scheme `dataname((:weight, mean))` you'll get "weight_mean". Hence, it breaks the assumption of argument order in the tests, so they now fail. The tests can be changed, but I wonder if it would make sense to swap the first and second entry like this
```julia
dataname(x::Tuple) = join(vcat([dataname(x[2]), dataname(x[1])], [dataname(s) for s in x[3:end]]), "_")
```
?

Reply:
> Sorry, missed that, just don't mention them in the docstring, that's enough.

Ok!

Reply:
Oh yea sorry forgot about that, yes do what you have written. This must absolutely be non-breaking! 

> The tests can be changed

No no, we have to keep it non breaking!

Reply:
Makes sense! However, removing the two- and three-argument versions of `dataname` breaks tests too (not to speak of the rename `aggname` -> `dataname`). Is it ok to alter the tests in those cases?

Reply:
Don't alter the tests at the moment. It'll be best to get the solution 'right', then we can revisit to see if that makes the most sense. Once everyone is agreed, we can alter tests as a final step: only if necessary.

Reply:
Ok, I've implemented now it as discussed (leaving out any alterations to tests), and rebased against latest master.